pedsnetdcc/db.py

```diff
@@ -0,0 +1,430 @@
+    statement. Able to execute the sql statement if given a connection string.
+
+    :raises RuntimeError: if setting the id_ attribute is attempted
+    """
+
+    logger = logger
+
+    def __init__(self, sql, msg='executing SQL', id_=None):
+        """Populate defaults on a new Statement object.
+
+        If msg is not passed, a default 'executing SQL' message is used. If id_
+        is not passed, a uuid is generated for the Statement.
+
+        :param str sql: the sql statement
+        :param str msg: a message describing the purpose of the sql
+        :param id_:     the unique identifer of the Statement
+        """
+        self.sql = sql
+        self.msg = msg
+        self._id_ = id_ or uuid.uuid4()
+        self.data = None
+        self.fields = []
+        self.rowcount = None
+        self.err = None
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+        return self.id_ == other.id_
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        """Return a unique int identifier; allows for StatementSet membership.
+
+        :returns: unique immutable identifier
+        :rtype:   int
+        """
+        return hash(self.id_)
+
+    def __repr__(self):
+        return 'Statement(sql={0}, msg={1}, id_={2})'.format(self.sql,
+                                                             self.msg,
+                                                             self.id_)
+
+    def __str__(self):
+        return 'Statement for {0}'.format(self.msg)
+
+    @property
+    def id_(self):
+        """Get the id_ value from its 'hidden' internal storage spot.
+
+        :returns: the id_ value
+        """
+        return self._id_
+
+    @id_.setter
+    def id_(self, val):
+        """Raise a RuntimeError if setting id_ is attempted.
+
+        :param val: a value to set id_ to
+        :raises:    RuntimeError on every invocation
+        """
+        raise RuntimeError('Statement.id_ is immutable.')
+
+    def execute(self, conn_str, resq=None, logq=None):
+        """Execute the sql statement against a database. Usable in parallel.
+
+        A new database connection is made using the passed connection string
+        and the `execute_on_conn` method is called with that connections and
+        `resq` and `logq`. After execution, regardless of errors, the
+        connection is closed.
+
+        :param str conn_str: connection string for the database
+        :param Queue resq:   result queue to pass to `execute_on_conn`
+        :param Queue logq:   logging queue to pass to `execute_on_conn`
+        :returns:            the processed Statement object itself
+        :rtype:              Statement
+        """
+
+        try:
+            with psycopg2.connect(conn_str) as conn:
+                self.execute_on_conn(conn, resq, logq)
+
+        finally:
+            conn.close()
```

Comment:
If there is a connection error, conn will not be defined, which will override whatever the original exception was.